feat: Introduce storage_client in DistributedRuntime

grahamking opened the pull request:
This will gradually replace `etcd_client`, and later extend to replacing nats_client. It is backed by etcd normally, or memory when etcd is not available.

Also use it to list instances, as an example and first usage.

This is the same interface that ModelDeploymentCards have always used for publishing and reading, so it is well tested.



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- New Features
  - Introduces a unified storage backend, enabling both in-memory and etcd-backed operation.
  - Adds a connection identifier for storage clients to aid diagnostics.
- Refactor
  - Migrates instance discovery and runtime services to the new storage abstraction for consistency across components.
  - Streamlines configuration: external coordination now requires an available storage backend.
- Bug Fixes
  - Improves resilience of health checks and instance listing by gracefully handling storage errors and continuing to serve responses with clear warnings.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

grahamking committed:
feat: Introduce storage_client in DistributedRuntime

This will gradually replace `etcd_client`. It is backed by etcd
normally, or memory when etcd is not available.

Also use it to list instances, as an example and first usage.

This is the same interface that ModelDeploymentCards have always used
for publishing and reading, so it is well tested.

Signed-off-by: Graham King <grahamk@nvidia.com>

grahamking committed:
Thanks Code Rabbit

Signed-off-by: Graham King <grahamk@nvidia.com>

nnshah1 reviewed (approved):
LGTM - but I think we should look at overall what services the runtime needs to provide - I'm not sure we need a distributed key value store - maybe its more that we need a way to store configuration and we need discovery? for model information we will use model express, for router - I'm thinking we would embed prefix tree storage into the router library, for discovery likewise we will use discovery interfaces... what remains in that case for key value store?

grahamking committed:
Rename things to progress code review

`*Storage` -> `*Store`
`storage_client` -> `store`
`etcd_root` -> `instance_root`

Add a comment explaining that although `store` is `dyn KeyValueStore`
right now, it has ambitions to be the one-ring of Dynamo storage.

Signed-off-by: Graham King <grahamk@nvidia.com>

keivenchang reviewed (commented):
I'm late to the game as this is already merged. Most of the other questions are already asked and answered. Concerns:

- Woule love to see comments explaining the reasons for some of the changes. For example, explain that Store now contains the same EtcdClient/etcd::Client because of xyz reasons (e.g. `// This is just a transition phase`)
- Naming. Thanks for making naming more consistent (e.g. Storage->Store). How about variable names, do you plan to rename variables like `etcd_client` (too specific) and `store` (too generic) to something like `keyval_store` in later PRs?
- Would love to see examples or documents on how we can replace KeyValueStore (currently EtcdClient) with memory instances in unit tests.
